### PR TITLE
Support latched topics in udp bridge

### PR DESCRIPTION
**Pull request by timonegk**
# Summary
* Adds support for latched / transient local topics by detecting them and republishing them periodically
* Removes base64 encoding of pickle before encryption
* Adds zlib compression before encryption to make it possible to receive larger messages (remark: upper limit due to network properties is 65535)
* Remove loop when receiving packets and remove packet delimiter because it is not necessary.
* Change info log to debug log

- [x] Run `colcon build`
- [ ] Write documentation
- [ ] Create issues for future work
- [x] Test on your machine
- [ ] Test on the robot
- [x] This PR is on our `Software` project board

### CODE REVIEW

File: udp_bridge/aes_helper.py [hunk 1]
```diff
@@ -35,12 +35,12 @@ def encrypt(self, message: str) -> bytes:
...
-        return Fernet(key=self.encryption_key).encrypt(bytes(message, encoding="UTF-8"))
+        return Fernet(key=self.encryption_key).encrypt(message)
 
     def decrypt(self, enc: bytes) -> str:
         if len(enc) == 0:
             raise ValueError("Cannot decrypt empty data")
         if self.encryption_key is None:
             return str(enc, encoding="UTF-8")
 
-        return str(Fernet(key=self.encryption_key).decrypt(enc), encoding="UTF-8")
```
Comment (lilioid) [inline udp_bridge/aes_helper.py]: As far as I'm aware the official documentation says that bytes are used as input to encrypt and output of decrypt.

https://cryptography.io/en/latest/fernet/#cryptography.fernet.Fernet.encrypt

**Review by lilioid — COMMENTED** [review]
the idea behind the message delimiter and the looping receive was that using that method larger messages than 65535 could be sent over the wire by splitting them into multiple udp packets.
This was never implemented on the sending side though :/

You could think about whether that is a useful property to have. If all relevant data packets including images are small enough then I agree that you won't need it.

**Review by texhnolyze — DISMISSED** [review]

File: udp_bridge/aes_helper.py (hunk 1 above)
Comment (timonegk) [reply udp_bridge/aes_helper.py]: Sorry, I forgot to update the type hints. `message` is now bytes.

**Comment by timonegk** [comment]
> the idea behind the message delimiter and the looping receive was that using that method larger messages than 65535 could be sent over the wire by splitting them into multiple udp packets. This was never implemented on the sending side though :/
> 
> You could think about whether that is a useful property to have. If all relevant data packets including images are small enough then I agree that you won't need it.

It would be nice to be able to do that, but the current state did not work for packets bigger than that size and even for packets smaller than that size, merging the packets did not actually work :/ (e.g. size 12000 bytes)

**Comment by timonegk** [comment]
Also, I am not sure if that can even work because packets might arrive in a different order.

**Comment by lilioid** [comment]
> Also, I am not sure if that can even work because packets might arrive in a different order.

yes they might; using udp is always on a best-effort basis.

> It would be nice to be able to do that, but the current state did not work for packets bigger than that size and even for packets smaller than that size, merging the packets did not actually work :/ (e.g. size 12000 bytes)

That might be because the sending side never actually did split the packets so they would just be chopped by the linux kernels udp handling.

**Comment by Flova** [comment]
Btw. maybe we should use the standard ROS serializer, instead of pickle. This might have security and compatibility reasons.

File: udp_bridge/sender.py [hunk 2]
```diff
@@ -55,19 +58,28 @@ def __subscribe(self, backoff=1.0):
...
+                        data_class,
+                        self.topic,
+                        lambda msg: self.__message_callback(msg, latched=True),
+                        QoSProfile(depth=1, durability=DurabilityPolicy.TRANSIENT_LOCAL)
+                )
+            self.node.get_logger().debug(f"Subscribed to topic {self.topic}")
         else:
             # topic is not yet known
-            self.node.get_logger().info(f"Topic {self.topic} is not yet known. Retrying in {backoff} seconds")
+            self.node.get_logger().debug(f"Topic {self.topic} is not yet known. Retrying in {backoff} seconds")
```
Comment (Flova) [inline udp_bridge/sender.py]: Maybe add a warning if the topic is not present after n seconds

**Review by Flova — DISMISSED** [review]
Some minor comments including running the auto formatter and addressing the logging

**Review by texhnolyze — APPROVED** [review]

**Pull request merged**